### PR TITLE
Refactor path config

### PR DESCRIPTION
**Pull request by kaklakariada**
Allow configuring a list of paths instead of pattern

### CODE REVIEW

File: src/main/java/org/itsallcode/openfasttrace/importer/legacytag/LegacyTagImporterFactory.java [hunk 1]
```diff
@@ -81,7 +83,7 @@ public Importer createImporter(final Path path, final Charset charset,
         if (!config.isPresent())
         {
             final List<String> patterns = this.config.get().getPathConfigs().stream() //
-                    .map(PathConfig::getPattern) //
+                    .map(PathConfig::getDescription) //
```
Comment (redcatbear) [inline src/main/java/org/itsallcode/openfasttrace/importer/legacytag/LegacyTagImporterFactory.java]: Variable name says "patterns" yet you see to be collecting "descriptions". Please unify.

File: src/main/java/org/itsallcode/openfasttrace/importer/legacytag/config/DescribedPathMatcher.java [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+        return new DescribedPathMatcher(paths.toString(), listMatcher);
+    }
+
+    private static String addMissingPatternPrefix(final String pattern)
+    {
+        if (pattern.startsWith(GLOB_PREFIX) || pattern.startsWith(REGEX_PREFIX))
+        {
+            return pattern;
+        }
+        return GLOB_PREFIX + pattern;
```
Comment (redcatbear) [inline src/main/java/org/itsallcode/openfasttrace/importer/legacytag/config/DescribedPathMatcher.java]: This default behavior should appear in both specification and design.

File: src/main/java/org/itsallcode/openfasttrace/importer/legacytag/config/PathConfig.java [hunk 3]
```diff
@@ -31,17 +32,13 @@
...
-    private final String pattern;
-    private final PathMatcher pathMatcher;
+    private final DescribedPathMatcher pathMatcher;
     private final String coveredItemNamePrefix;
     private final String coveredItemArtifactType;
     private final String tagArtifactType;
 
     /**
-     * Create a new path configuration.
+     * Create a new pattern based path configuration.
```
Comment (redcatbear) [inline src/main/java/org/itsallcode/openfasttrace/importer/legacytag/config/PathConfig.java]: "pattern-based"

File: src/main/java/org/itsallcode/openfasttrace/importer/legacytag/config/PathConfig.java [hunk 4]
```diff
@@ -56,44 +53,29 @@
      * @param tagArtifactType
      *            the artifact type of the imported {@link SpecificationItem}s.
      */
-    public PathConfig(final String pattern, final String coveredItemArtifactType,
-            final String coveredItemNamePrefix, final String tagArtifactType)
+    public static PathConfig createPatternConfig(final String pattern,
```
Comment (redcatbear) [inline src/main/java/org/itsallcode/openfasttrace/importer/legacytag/config/PathConfig.java]: Four unnamed arguments. Replace factory method by builder?

File: src/main/java/org/itsallcode/openfasttrace/importer/legacytag/config/PathConfig.java [hunk 5]
```diff
@@ -116,9 +98,8 @@ public String getCoveredItemNamePrefix()
     @Override
     public String toString()
     {
-        return "PathConfig [pattern=" + this.pattern + ", pathMatcher=" + this.pathMatcher
-                + ", coveredItemNamePrefix=" + this.coveredItemNamePrefix
-                + ", coveredItemArtifactType=" + this.coveredItemArtifactType + ", tagArtifactType="
-                + this.tagArtifactType + "]";
+        return "PathConfig [pathMatcher=" + this.pathMatcher.toString() + ", coveredItemNamePrefix="
```
Comment (redcatbear) [inline src/main/java/org/itsallcode/openfasttrace/importer/legacytag/config/PathConfig.java]: Maybe enclose the comma separated parts in quotes for better readability of the output?

File: src/test/java/org/itsallcode/openfasttrace/importer/legacytag/TestLegacyTagImporterFactory.java [hunk 6]
```diff
@@ -149,6 +151,6 @@ private LegacyTagImporterFactory create(final LegacyTagImporterConfig config)
 
     private PathConfig glob(final String globPattern)
     {
-        return new PathConfig("glob:" + globPattern, null, null, null);
+        return PathConfig.createPatternConfig("glob:" + globPattern, null, null, null);
```
Comment (redcatbear) [inline src/test/java/org/itsallcode/openfasttrace/importer/legacytag/TestLegacyTagImporterFactory.java]: Yep, definitely replace by a builder. This is too hard too read.

File: src/test/java/org/itsallcode/openfasttrace/importer/legacytag/config/DescribedPathMatcherTest.java [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+    private void assertMatches(final String path, final boolean expected)
+    {
+        assertThat("path " + path, this.matcher.matches(Paths.get(path)), equalTo(expected));
+    }
+
+    private void assertDescription(final String expected)
+    {
+        assertThat(this.matcher.getDescription(), equalTo(expected));
+    }
+}
```
Comment (redcatbear) [inline src/test/java/org/itsallcode/openfasttrace/importer/legacytag/config/DescribedPathMatcherTest.java]: Add tests for Windows-style paths too. Just to be safe.

File: src/main/java/org/itsallcode/openfasttrace/importer/legacytag/config/DescribedPathMatcher.java [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+    }
+
+    String getDescription()
+    {
+        return this.description;
+    }
+
+    boolean matches(final Path path)
```
Comment (redcatbear) [inline src/main/java/org/itsallcode/openfasttrace/importer/legacytag/config/DescribedPathMatcher.java]: package scope on purpose?

File: src/main/java/org/itsallcode/openfasttrace/importer/legacytag/config/DescribedPathMatcher.java [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+    private final String description;
+    private final PathMatcher matcher;
+
+    private DescribedPathMatcher(final String description, final PathMatcher matcher)
+    {
+        this.description = description;
+        this.matcher = matcher;
+    }
+
+    static DescribedPathMatcher createPatternMatcher(final String pattern)
```
Comment (redcatbear) [inline src/main/java/org/itsallcode/openfasttrace/importer/legacytag/config/DescribedPathMatcher.java]: Package scope on purpose?

File: src/main/java/org/itsallcode/openfasttrace/importer/legacytag/config/DescribedPathMatcher.java [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+    {
+        final String fullPattern = addMissingPatternPrefix(pattern);
+        final PathMatcher patternMatcher = FileSystems.getDefault().getPathMatcher(fullPattern);
+        return new DescribedPathMatcher(fullPattern, patternMatcher);
+    }
+
+    static DescribedPathMatcher createPathListMatcher(final List<Path> paths)
```
Comment (redcatbear) [inline src/main/java/org/itsallcode/openfasttrace/importer/legacytag/config/DescribedPathMatcher.java]: Package scope on purpose?

File: src/main/java/org/itsallcode/openfasttrace/importer/legacytag/config/DescribedPathMatcher.java [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+    private static String addMissingPatternPrefix(final String pattern)
+    {
+        if (pattern.startsWith(GLOB_PREFIX) || pattern.startsWith(REGEX_PREFIX))
+        {
+            return pattern;
+        }
+        return GLOB_PREFIX + pattern;
+    }
+
+    String getDescription()
```
Comment (redcatbear) [inline src/main/java/org/itsallcode/openfasttrace/importer/legacytag/config/DescribedPathMatcher.java]: Package scope on purpose?

File: src/main/java/org/itsallcode/openfasttrace/importer/legacytag/config/DescribedPathMatcher.java [hunk 12]
```diff
@@ -0,0 +1,93 @@
...
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ * 
+ * You should have received a copy of the GNU General Public
+ * License along with this program.  If not, see
+ * <http://www.gnu.org/licenses/gpl-3.0.html>.
+ * #L%
+ */
+
```
Comment (redcatbear) [inline src/main/java/org/itsallcode/openfasttrace/importer/legacytag/config/DescribedPathMatcher.java]: Please add JavaDoc to all public methods. To package-scoped optionally - but recommended.

**Comment by redcatbear** [comment]
Please document the whole legacy format and matchers in spec and design document.

File: src/main/java/org/itsallcode/openfasttrace/importer/legacytag/LegacyTagImporterFactory.java (hunk 1 above)
Comment (kaklakariada) [reply src/main/java/org/itsallcode/openfasttrace/importer/legacytag/LegacyTagImporterFactory.java]: :heavy_check_mark:

File: src/main/java/org/itsallcode/openfasttrace/importer/legacytag/config/PathConfig.java (hunk 3 above)
Comment (kaklakariada) [reply src/main/java/org/itsallcode/openfasttrace/importer/legacytag/config/PathConfig.java]: :heavy_check_mark:

File: src/main/java/org/itsallcode/openfasttrace/importer/legacytag/config/PathConfig.java (hunk 5 above)
Comment (kaklakariada) [reply src/main/java/org/itsallcode/openfasttrace/importer/legacytag/config/PathConfig.java]: :heavy_check_mark:

File: src/test/java/org/itsallcode/openfasttrace/importer/legacytag/config/DescribedPathMatcherTest.java (hunk 7 above)
Comment (kaklakariada) [reply src/test/java/org/itsallcode/openfasttrace/importer/legacytag/config/DescribedPathMatcherTest.java]: :heavy_check_mark:

File: src/main/java/org/itsallcode/openfasttrace/importer/legacytag/config/DescribedPathMatcher.java (hunk 8 above)
Comment (kaklakariada) [reply src/main/java/org/itsallcode/openfasttrace/importer/legacytag/config/DescribedPathMatcher.java]: Yes. `DescribedPathMatcher` is only used in this package.

File: src/main/java/org/itsallcode/openfasttrace/importer/legacytag/config/DescribedPathMatcher.java (hunk 12 above)
Comment (kaklakariada) [reply src/main/java/org/itsallcode/openfasttrace/importer/legacytag/config/DescribedPathMatcher.java]: This is only used internally. Users only work with `PathConfig`.

File: src/main/java/org/itsallcode/openfasttrace/importer/legacytag/config/PathConfig.java (hunk 4 above)
Comment (kaklakariada) [reply src/main/java/org/itsallcode/openfasttrace/importer/legacytag/config/PathConfig.java]: :heavy_check_mark:

File: src/main/java/org/itsallcode/openfasttrace/importer/legacytag/config/DescribedPathMatcher.java (hunk 8 above)
Comment (redcatbear) [reply src/main/java/org/itsallcode/openfasttrace/importer/legacytag/config/DescribedPathMatcher.java]: :white_check_mark:

File: src/main/java/org/itsallcode/openfasttrace/importer/legacytag/config/DescribedPathMatcher.java (hunk 12 above)
Comment (redcatbear) [reply src/main/java/org/itsallcode/openfasttrace/importer/legacytag/config/DescribedPathMatcher.java]: :heavy_check_mark:

**Pull request merged**